java-client/src/test/java/org/opensearch/client/opensearch/integTest/AbstractClusterClientIT.java

```diff
@@ -93,10 +93,17 @@ public void testClusterUpdateSettingNonExistent() throws IOException {
         try {
             openSearchClient.cluster().putSettings(request);
             fail();
-        } catch (OpenSearchException e) {
-            assertNotNull(e);
-            assertEquals(e.response().status(), 400);
-            assertTrue(e.getMessage().contains("transient setting [no_idea_what_you_are_talking_about], not recognized"));
+        } catch (Exception e) {
+            OpenSearchException openSearchException;
+            if (e instanceof OpenSearchException) {
```

Comment:
I wouldn't expect conditionals in tests, is this not predictable?

Reply:
It's predictable but now varies per implementation.

This test is run against the RestClient transport and the Apache HTTP 5 transports. The RestClient transport throws OpenSearchException directly and since its transport is synchronous, this exception's stack trace contains a reference to the caller. The HC5 transport is _always_ asynchronous and while the underlying logic will throw an OpenSearchException, that logic is executing on a pooled thread. That OpenSearchException then has to be wrapped in order to preserve a reference to the caller in the thrown exception's stack trace. I cannot construct an OpenSearchException wrapping an OpenSearchException (which would be suspicious anyways), so this now has to catch any exception.

The correct way to model this is to parameterize each _implementation_ of this abstract class by the type of exception it should be expected to throw and use the `assertThrows` pattern here instead of `try-catch`.

This modeling issue combined with https://github.com/opensearch-project/opensearch-java/pull/656#discussion_r1349301994 are two overarching pieces of test tech debt I think should be handled separately from this PR, but would enable more comprehensive transport-agnostic integration tests.

Reply:
Thanks for the detailed explanation. Open GitHub issues for these leftovers so we don't forget / someone else can help.